### ISSUE TITLE
new major version of json gem breaks default plugin installation rake task

### ISSUE THREAD

**Issue by jsvd** [elastic/logstash#5648]
because the only gems that depend on json during rake bootstrap + test:install-core don't force `gem "json", "~> 1"`, when installing the default plugins we get the following:

```
/tmp/logstash (git)-[2.4] % rake artifact:tar
mkdir -p vendor
mkdir vendor/_
Downloading http://jruby.org.s3.amazonaws.com/downloads/1.7.25/jruby-bin-1.7.25.tar.gz
Installing minitar >= 0 because the build process needs it.
[bootstrap] Fetching and installing gem: minitar (>= 0)
Fetching: minitar-0.5.4.gem (100%)
Successfully installed minitar-0.5.4
Installing bundler ~> 1.9.4 because the build process needs it.
[bootstrap] Fetching and installing gem: bundler (~> 1.9.4)
Fetching: bundler-1.9.10.gem (100%)
Successfully installed bundler-1.9.10
Invoking bundler install...
Fetching gem metadata from https://rubygems.org/.........
Fetching version metadata from https://rubygems.org/...
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies....
Installing addressable 2.3.8
Installing cabin 0.8.1
Installing arr-pm 0.0.10
Installing backports 3.6.8
Installing ffi 1.9.14
Installing childprocess 0.5.9
Installing clamp 0.6.5
Installing coderay 1.1.1
Installing concurrent-ruby 0.9.2
Installing multipart-post 2.0.0
Installing faraday 0.9.2
Using minitar 0.5.4
Installing file-dependencies 0.1.6
Installing filesize 0.0.4
Installing json 2.0.1
Installing fpm 1.3.3
Installing gems 0.8.3
Installing i18n 0.6.9
Installing jrjackson 0.3.9
Using jruby-openssl 0.9.16
Using logstash-core-event 2.4.0.dev from source at ./logstash-core-event
Installing method_source 0.8.2
Installing slop 3.6.0
Installing spoon 0.0.4
Installing pry 0.10.4
Installing rubyzip 1.1.7
Installing stud 0.0.22
Installing thread_safe 0.3.5
Installing polyglot 0.3.5
Installing treetop 1.4.15
Using logstash-core 2.4.0.dev from source at ./logstash-core
Using logstash-core-plugin-api 1.60.0 from source at ./logstash-core-plugin-api
Installing sawyer 0.6.0
Installing octokit 3.8.0
Using bundler 1.9.10
Bundle complete! 17 Gemfile dependencies, 35 gems now installed.
Gems in the group development were not installed.
Bundled gems are installed into ./vendor/bundle.
Building logstash-core-event-java using gradle
:compileJava UP-TO-DATE
:processResources UP-TO-DATE
:classes UP-TO-DATE
:jar UP-TO-DATE
:sourcesJar UP-TO-DATE
:copyGemjar

BUILD SUCCESSFUL

Total time: 4.557 secs

This build could be faster, please consider using the Gradle Daemon: https://docs.gradle.org/2.8/userguide/gradle_daemon.html
[plugin:install-default] Installing default plugins
Installing logstash-input-heartbeat, logstash-output-zeromq, logstash-codec-collectd, logstash-output-xmpp, logstash-codec-dots, logstash-codec-edn, logstash-codec-edn_lines, logstash-codec-fluent, logstash-codec-es_bulk, logstash-codec-graphite, logstash-codec-json, logstash-codec-json_lines, logstash-codec-line, logstash-codec-msgpack, logstash-codec-multiline, logstash-codec-netflow, logstash-codec-oldlogstashjson, logstash-codec-plain, logstash-codec-rubydebug, logstash-filter-anonymize, logstash-filter-checksum, logstash-filter-clone, logstash-filter-csv, logstash-filter-date, logstash-filter-dns, logstash-filter-drop, logstash-filter-fingerprint, logstash-filter-geoip, logstash-filter-grok, logstash-filter-json, logstash-filter-kv, logstash-filter-metrics, logstash-filter-multiline, logstash-filter-mutate, logstash-filter-ruby, logstash-filter-sleep, logstash-filter-split, logstash-filter-syslog_pri, logstash-filter-throttle, logstash-filter-urldecode, logstash-filter-useragent, logstash-filter-uuid, logstash-filter-xml, logstash-input-couchdb_changes, logstash-input-elasticsearch, logstash-input-eventlog, logstash-input-exec, logstash-input-file, logstash-input-ganglia, logstash-input-gelf, logstash-input-generator, logstash-input-graphite, logstash-input-http, logstash-input-http_poller, logstash-input-imap, logstash-input-irc, logstash-input-jdbc, logstash-input-log4j, logstash-input-lumberjack, logstash-input-pipe, logstash-input-rabbitmq, logstash-input-redis, logstash-input-s3, logstash-input-snmptrap, logstash-input-sqs, logstash-input-stdin, logstash-input-syslog, logstash-input-tcp, logstash-input-twitter, logstash-input-udp, logstash-input-unix, logstash-input-xmpp, logstash-input-zeromq, logstash-input-kafka, logstash-input-beats, logstash-output-cloudwatch, logstash-output-csv, logstash-output-elasticsearch, logstash-output-email, logstash-output-exec, logstash-output-file, logstash-output-ganglia, logstash-output-gelf, logstash-output-graphite, logstash-output-hipchat, logstash-output-http, logstash-output-irc, logstash-output-juggernaut, logstash-output-lumberjack, logstash-output-nagios, logstash-output-nagios_nsca, logstash-output-null, logstash-output-opentsdb, logstash-output-pagerduty, logstash-output-pipe, logstash-output-rabbitmq, logstash-output-redis, logstash-output-s3, logstash-output-sns, logstash-output-sqs, logstash-output-statsd, logstash-output-stdout, logstash-output-tcp, logstash-output-udp, logstash-output-kafka
Plugin version conflict, aborting
ERROR: Installation Aborted, message: Bundler could not find compatible versions for gem "logstash-core-plugin-api":
  In snapshot (Gemfile.lock):
    logstash-core-plugin-api (= 1.60.0)

  In Gemfile:
    logstash-output-sqs (>= 0) java depends on
      logstash-core-plugin-api (~> 1.0) java

    logstash-output-sqs (>= 0) java depends on
      logstash-core-plugin-api (~> 1.0) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-output-sqs (>= 0) java depends on
      logstash-core-plugin-api (~> 1.0) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-output-sqs (>= 0) java depends on
      logstash-core-plugin-api (~> 1.0) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-output-sqs (>= 0) java depends on
      logstash-core-plugin-api (~> 1.0) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-pipe (>= 0) java depends on
      logstash-core-plugin-api (<= 2.99, >= 1.60) java

    logstash-input-sqs (>= 0) java depends on
      logstash-mixin-aws (>= 1.0.0) java depends on
        logstash-core-plugin-api (~> 2.0) java

    logstash-core-plugin-api (= 1.60.0) java

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
Bundler could not find compatible versions for gem "json":
  In snapshot (Gemfile.lock):
    json (= 2.0.1)

  In Gemfile:
    simplecov (>= 0) java depends on
      json (< 3, >= 1.8) java

    simplecov (>= 0) java depends on
      json (< 3, >= 1.8) java

    fpm (~> 1.3.3) java depends on
      json (>= 1.7.7) java

    logstash-input-couchdb_changes (>= 0) java depends on
      json (>= 0) java

    logstash-input-sqs (>= 0) java depends on
      logstash-mixin-aws (>= 1.0.0) java depends on
        aws-sdk-v1 (>= 1.61.0) java depends on
          json (~> 1.4) java

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
Bundler could not find compatible versions for gem "logstash-core":
  In snapshot (Gemfile.lock):
    logstash-core (= 2.4.0.dev)

  In Gemfile:
    logstash-core-plugin-api (= 1.60.0) java depends on
      logstash-core (<= 2.4.99, >= 2.4.0.alpha) java

    logstash-output-zeromq (>= 0) java depends on
      logstash-core (< 3.0.0, >= 2.0.0.beta2) java

    logstash-output-zeromq (>= 0) java depends on
      logstash-mixin-zeromq (>= 2.0.0) java depends on
        logstash-core (< 3.0.0, >= 2.0.0) java

    logstash-input-sqs (>= 0) java depends on
      logstash-core (< 2.0.0, >= 1.4.0) java

    logstash-core (= 2.4.0.dev) java

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
Bundler could not find compatible versions for gem "logstash":
  In Gemfile:
    logstash-input-sqs (>= 0) java depends on
      logstash (< 2.0.0, >= 1.4.0) java
Could not find gem 'logstash (< 2.0.0, >= 1.4.0) java', which is required by gem 'logstash-input-sqs (>= 0) java', in any of the sources.
rake aborted!

/private/tmp/logstash/rakelib/plugin.rake:7:in `install_plugins'
/private/tmp/logstash/rakelib/plugin.rake:27:in `(root)'
org/jruby/RubyProc.java:281:in `call'
org/jruby/RubyArray.java:1613:in `each'
org/jruby/RubyArray.java:1613:in `each'
org/jruby/RubyArray.java:1613:in `each'
org/jruby/RubyArray.java:1613:in `each'
org/jruby/RubyKernel.java:1059:in `load'
Tasks: TOP => artifact:tar => artifact:prepare => plugin:install-default
(See full trace by running task with --trace)
```

**Comment by jsvd** [comment]
The main reason json v2.x.x gets installed is from the json dependencies of:

* coveralls: `json < 3, >= 1.8`
* simplecov: `json < 3, >= 1.8`
* fpm 1.3.x:  `json >= 1.7.7`

Looking at the latest version of fpm, the json dependency has been limited to:

* fpm 1.6.1: `json >= 1.7.7`
* fpm 1.6.2: `json < 2.0, >= 1.7.7`

This means we have two solutions here:

1) pin `gem "json", "~> 1"` in the Gemfile
2) upgrade the fpm dependency in the Gemfile to `gem "fpm", "~> 1.6.3"` (equivalent to 1.6.x but bigger than or equal to 1.6.3)

Is there any issue preventing 2), @jordansissel ?

**Comment by suyograo** [comment]
@jordansissel and I had a zoom conversation 'bout this. Looks like the culprit is:

```
 logstash-input-sqs (>= 0) java depends on
      logstash-mixin-aws (>= 1.0.0) java depends on
        aws-sdk-v1 (>= 1.61.0) java depends on
          json (~> 1.4) java
```

See `json (~> 1.4)` which is not allowing json 2.0 to be installed. How we got to 2.0 in `*.lock` file is a wonder.. We should probably delete and regenerate the 2.4 lock file..

```
Bundler could not find compatible versions for gem "json":
  In snapshot (Gemfile.lock):
    json (= 2.0.1)
```

**Comment by jsvd** [comment]
We've decided to go ahead with:

2) upgrade the fpm dependency in the Gemfile to gem "fpm", "~> 1.6.3" (equivalent to 1.6.x + bigger than or equal to 1.6.3)

we'll do this and check if packages are ok (they should be as this is a minor version bump from fpm 1.3 to 1.6)

**Comment by suyograo** [comment]
Just to be clear, `fpm` is installed first which locks the `json` dependency to 1.x which is then ok with dependency requirements from other gems downstream (like sqs). Long term, when we move to `aws-sdk` v2, the json dependency constraint will be removed

**Comment by suyograo** [comment]
This is seemingly fixed in 2.4 and packages are building fine http://build-eu-00.elastic.co/view/LS%202.4/job/logstash_regression_24/changes
